tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,173 @@
+
+``` 
+describe "id" $
+     it "generates the right output with the right params" $
+        let output = show $ id (1:: Int)
+        in defaultGolden "id" output
+```
+
+On this test we are setting up a B.D.D. test with [hspec](https://github.com/hspec/hspec) “describe” and “it”,
+but on the inside there is an implementation of the `defaultGolden` helper,
+that will create(or update) for us a directory called `.golden` this will contain another directory named after our described test in this case `id` with two files containing the `actual` and `golden` values for our test.
+
+For our example we will create our own Golden type.
+We have created a file called test.yml that contains an ansible role on the dir ["test/fixtures/"](https://github.com/MedAdrian/hspec-golden-tutorial/blob/master/test/fixtures/test.yml) 
+we will use this file contents and transform them into JSON.
+
+```
+{--
+Here we define our Golden test as follows:
+- We read a .yml file that contains an ansible role
+- We decode the string read by Haskell that we read form the .yml file
+--}
+
+spec :: Spec
+spec = do
+    describe "conversion" $ do
+      content <- runIO $ B.readFile "test/fixtures/test.yml"
+      let eYaml = Y.decodeEither' content :: Either Y.ParseException Y.Value  --Parses to a generic yml type.
+          Right json = A.encode <$> eYaml                                     --Parses the data type to a json structure.
+      it "transforms a YAML file contents to JSON" $ do
+        myGoldenTest "conversion" json                                        --Generates the golden test.
+```
+
+To see the full code example access this [repository](https://github.com/MedAdrian/hspec-golden-tutorial).
+
+##### I.e: Directory Structure 
+``` 
+|_.golden
+   |_id
+     |_actual
+     |_golden
+```
+
+For this example we are testing that the output from the function will return the expected golden value.
+
+## Installing the CLI
+You can install the [hspec-golden](https://github.com/stackbuilders/hspec-golden) command line interface (CLI) with stack:
+
+``` $ stack install hspec-golden ```
+
+or cabal:
+
+``` $ cabal install hspec-golden ```
+
+Update the golden tests under .golden directory:
+
+``` $ hgold -u ```
+
+Update the golden tests under .myGoldenTest directory:
+
+``` $ hgold -u .myGoldenTest ```
+
+## Conclusions
+We can determine that the name “golden file” only refers to the file with the output value, not the input.
+On most libraries there is no input values stored in a file,
+but in practice is often convenient to store them both in files so that there is an input file for every output file and vice versa.
+
+A golden test can fail when the actual output value does not match the contents of the golden file.
+You then need to determine whether this is an intentional code change or a bug.
```

Comment:
This part should be one of the examples above. That is, how does the output look when there's an error? Or first when there's a success and then if you break something.

Reply:
You are right, we are missing the cases where this fails.